3_visualize/src/svg_utils_mapping.R

```diff
@@ -0,0 +1,167 @@
+  }
+  
+  usa_sf <- maps::map("usa", fill = TRUE, plot=FALSE) %>%
+    sf::st_as_sf() %>% 
+    st_transform(proj_str) %>% 
+    st_buffer(0) 
+  
+  if(outline_states) usa_sf <- use_state_outlines(usa_sf, proj_str)
+  
+  return(usa_sf)
+}
+
+use_state_outlines <- function(usa_border_sf, proj_str) {
+  
+  # Need to remove islands from state outlines and then add back in 
+  # later so that they can be drawn as individual polygons. Otherwise,
+  # drawn with the state since the original state maps data only has 1
+  # ID per state.
+  
+  usa_islands_sf <- usa_border_sf %>% filter(ID != "main")
+  usa_addl_islands_sf <- generate_addl_islands(proj_str)
+  usa_mainland_sf <- usa_border_sf %>% 
+    filter(ID == "main") %>% 
+    st_erase(usa_addl_islands_sf) 
+  
+  # Have to manually add in CO because in `maps`, it is an incomplete
+  # polygon and gets dropped somewhere along the way.
+  co_sf <- maps::map("state", "colorado", fill = TRUE, plot=FALSE) %>%
+    sf::st_as_sf() %>%
+    st_transform(proj_str)
+  
+  maps::map("state", fill = TRUE, plot=FALSE) %>%
+    sf::st_as_sf() %>%
+    st_transform(proj_str) %>%
+    st_buffer(0) %>% 
+    # Get rid of islands from state outline data
+    st_intersection(usa_mainland_sf) %>%
+    select(-ID.1) %>% # st_intersection artifact that is unneeded
+    # Add islands back in as separate polygons from states
+    bind_rows(usa_islands_sf) %>%
+    bind_rows(usa_addl_islands_sf) %>% 
+    st_buffer(0) %>%
+    st_cast("MULTIPOLYGON") %>% # Needed to bring back to correct type to use st_coordinates
+    rmapshaper::ms_simplify(0.5) %>%
+    bind_rows(co_sf) # bind CO after bc otherwise it gets dropped in st_buffer(0)
+  
+}
+
+generate_addl_islands <- function(proj_str) {
+  # These are not called out specifically as islands in the maps::map("usa") data
+  # but cause lines to be drawn across the map if not treated separately. This creates those shapes.
+  
+  # Counties to be considered as separate polygons
+  
+  separate_polygons <- list(
+    `upper penninsula` = list(
+      state = "michigan",
+      counties = c(
+        "alger",
+        "baraga",
+        "chippewa",
+        "delta",
+        "dickinson",
+        "gogebic",
+        "houghton",
+        "iron",
+        "keweenaw",
+        "luce",
+        "mackinac",
+        "marquette",
+        "menominee",
+        "ontonagon",
+        "schoolcraft"
+      )),
+    `eastern shore` = list(
+      state = "virginia",
+      counties = c(
+        "accomack",
+        "northampton"
+      )),
+    # TODO: borders still slightly wonky bc it doesn't line up with counties perfectly. 
+    `nags head` = list(
+      state = "north carolina",
+      counties = c(
+        "currituck"
+      )),
+    # This + simplifying to 0.5 took care of the weird line across NY
+    `staten island` = list(
+      state = "new york",
+      counties = c(
```

Comment:
Interesting. I've found success with CO also by first simplifying the polygons with `ms_simplify()` and then `st_buffer(0)`. I think the simplification makes the coordinates slightly more distinct where they don't intersect, and therefore respect their shapes a little better

Reply:
This is code I borrowed from the WBEEP water use map I made. I am using the spatial data included in the `maps` package, which is not exactly perfect. It gets dropped after applying the Albers Equal Area transform ... which is so weird.

```r
library(sf)
library(dplyr)

# It's there:
yes_co_sf <- maps::map("state", fill = TRUE, plot=FALSE) %>%
    sf::st_as_sf() %>%
    st_buffer(0)
plot(st_geometry(yes_co_sf ), col = "grey")

# Now it's not:
 proj_str <- "+proj=aea +lat_1=29.5 +lat_2=45.5 +lat_0=37.5 +lon_0=-96 +x_0=0 +y_0=0 +ellps=GRS80 +datum=NAD83 +units=m +no_defs"
no_co_sf <- maps::map("state", fill = TRUE, plot=FALSE) %>%
    sf::st_as_sf() %>%
    st_transform(proj_str) %>%
    st_buffer(0)
plot(st_geometry(no_co_sf), col = "grey")
```

Reply:
Pinpointing the problem a bit more ... it appears to only happen when I do `st_buffer(0)` AFTER transforming. If left out or done before the transform, Colorado sticks around.

```r
# Buffer after transform = no Colorado!
no_co_sf <- maps::map("state", fill = TRUE, plot=FALSE) %>%
    sf::st_as_sf() %>%
    st_transform(proj_str) %>% 
    st_buffer(0)
plot(st_geometry(no_co_sf), col = "grey")

# Buffer before transform = yay! there's CO
yay_co_sf <- maps::map("state", fill = TRUE, plot=FALSE) %>%
    sf::st_as_sf() %>% 
    st_buffer(0) %>%
    st_transform(proj_str)
plot(st_geometry(yay_co_sf), col = "grey")

# No buffer at all - yes, we still have CO!
yes_co_sf <- maps::map("state", fill = TRUE, plot=FALSE) %>%
    sf::st_as_sf() %>% 
    st_transform(proj_str)
plot(st_geometry(yes_co_sf), col = "grey")
```

Reply:
In the interest of time, I'm going to stick this info into an issue to circle back later.

Reply:
See #23